src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/Reliability/Common/SqlRetryLogic.cs

```diff
@@ -0,0 +1,104 @@
+        private const int maxAttempts = 60;
+
+        private readonly string _typeName = nameof(SqlRetryLogic);
+
+        public Predicate<string> PreCondition { get; private set; }
+
+        public SqlRetryLogic(int numberOfTries,
+                             SqlRetryIntervalBaseEnumerator enumerator,
+                             Predicate<Exception> transientPredicate,
+                             Predicate<string> preCondition)
+        {
+            Debug.Assert(enumerator != null, $"The '{nameof(enumerator)}' mustn't be null.");
+            Debug.Assert(transientPredicate != null, $"The '{nameof(transientPredicate)}' mustn't be null.");
+
+            if (!(numberOfTries > counterDefaultValue && numberOfTries <= maxAttempts))
+            {
+                // The 'numberOfTries' should be between 1 and 60.
+                throw new ArgumentOutOfRangeException(nameof(numberOfTries), StringsHelper.GetString(Strings.SqlRetryLogic_InvalidRange, numberOfTries, counterDefaultValue + 1, maxAttempts));
+            }
+
+            NumberOfTries = numberOfTries;
+            RetryIntervalEnumerator = enumerator;
+            TransientPredicate = transientPredicate;
+            PreCondition = preCondition;
+            Current = counterDefaultValue;
+        }
+
+        public SqlRetryLogic(int numberOfTries, SqlRetryIntervalBaseEnumerator enumerator, Predicate<Exception> transientPredicate)
+            : this(numberOfTries, enumerator, transientPredicate, null)
+        {
+        }
+
+        public SqlRetryLogic(SqlRetryIntervalBaseEnumerator enumerator, Predicate<Exception> transientPredicate)
+            : this(counterDefaultValue + 1, enumerator, transientPredicate)
+        {
+        }
+
+        public override void Reset()
+        {
+            Current = counterDefaultValue;
+            RetryIntervalEnumerator.Reset();
+        }
+
+        public override bool TryNextInterval(out TimeSpan intervalTime)
+        {
+            intervalTime = TimeSpan.Zero;
+            // First try has occurred before starting the retry process. 
+            bool result = Current < NumberOfTries - 1;
+
+            if (result)
+            {
+                Current++;
+                // it doesn't mind if the enumerator gets to the last value till the number of attempts ends.
+                RetryIntervalEnumerator.MoveNext();
+                intervalTime = RetryIntervalEnumerator.Current;
+                SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> Next gap time will be '{2}' before the next retry number {3}", _typeName, MethodBase.GetCurrentMethod().Name, intervalTime, Current);
+            }
+            else
+            {
+                SqlClientEventSource.Log.TryTraceEvent("<sc.{0}.{1}|INFO> Current retry ({2}) has reached to the maximum attempts (total attempts except the first run = {3}).", _typeName, MethodBase.GetCurrentMethod().Name, Current, NumberOfTries - 1);
```

Comment:
Do we not throw an exception here?

Reply:
It could be a solution. But I'd rather handle it by a boolean instead of throwing an exception from the performance perspective.